rofl-scheduler/src/serverd/auth.rs

```diff
@@ -0,0 +1,234 @@
+use std::sync::{Arc, LazyLock};
+
+use anyhow::{anyhow, Context};
+use axum::{extract, http, response, RequestPartsExt};
+use axum_extra::{
+    headers::{authorization::Bearer, Authorization},
+    TypedHeader,
+};
+use base64::{prelude::BASE64_STANDARD, Engine};
+use chrono::{prelude::*, Duration};
+use rand::RngCore;
+use rustc_hex::FromHex;
+
+use oasis_runtime_sdk::{
+    crypto::signature::{self, PublicKey},
+    types::address::{Address, SignatureAddressSpec},
+};
+
+use super::{error::Error, State};
+
+/// Global JWT keys instance.
+static JWT_KEYS: LazyLock<Keys> = LazyLock::new(Keys::generate);
+
+/// JWT keys.
+struct Keys {
+    encoding: jsonwebtoken::EncodingKey,
+    decoding: jsonwebtoken::DecodingKey,
+}
+
+impl Keys {
+    /// Generate a new pair of JWT keys.
+    fn generate() -> Self {
+        let mut key = [0; 32];
+        rand::rngs::OsRng.fill_bytes(&mut key);
+
+        Self {
+            encoding: jsonwebtoken::EncodingKey::from_secret(&key),
+            decoding: jsonwebtoken::DecodingKey::from_secret(&key),
+        }
+    }
+}
+
+/// Claims issued after successful authentication.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct Claims {
+    pub address: String,
+    exp: u64,
+}
+
+impl<S> extract::FromRequestParts<S> for Claims
+where
+    S: Send + Sync,
+{
+    type Rejection = Error;
+
+    async fn from_request_parts(
+        parts: &mut http::request::Parts,
+        _state: &S,
+    ) -> Result<Self, Self::Rejection> {
+        // Extract the token from the authorization header.
+        let TypedHeader(Authorization(bearer)) = parts
+            .extract::<TypedHeader<Authorization<Bearer>>>()
+            .await
+            .map_err(|_| Error::BadAuthToken)?;
+
+        // Decode the user data.
+        let token_data = jsonwebtoken::decode::<Claims>(
+            bearer.token(),
+            &JWT_KEYS.decoding,
+            &jsonwebtoken::Validation::default(),
+        )
+        .map_err(|_| Error::BadAuthToken)?;
+
+        Ok(token_data.claims)
+    }
+}
+
+/// Login request.
+#[derive(Debug, Clone, serde::Deserialize)]
+#[serde(tag = "method", content = "data")]
+pub enum AuthLoginRequest {
```

Comment:
No nonce usage? I gues that would complicate things a bit, as the API would need to provision and temporarily store them? 

For this use-case just relying on expiry might be good enough.

Reply:
There are several ways to complicate things with the nonce:
* Introduce a two-step process where the server gets you the nonce that you need to use. This requires an additional API call. You could make it stateless (so it doesn't consume memory) by computing the nonce using `Timestamp || HMAC[NonceKey](Address || Timestamp)` and require that the Timestamp is at most X seconds old. Or just use a small cache.
* Use the on-chain account nonce in some way, so e.g. signing a transaction would invalidate all sign-in messages.

Thoughts?

Reply:
> The stateless variant is intriguing, but I don’t think it offers significantly stronger security guarantees

No, it is just about reducing memory usage from the need to store nonces and instead storing the nonce at the client (similar as what JWTs are doing so you don't need to keep state on the server). It also avoids the problem with someone polluting the (small) cache and causing legitimate requests to be evicted before they can be completed.

> requiring a short expiry (not_after field),

The current implementation makes it so that `not_after` (if lower than 6 hours) is respected in the issued JWT, so the whole session then has a shorter validity.

Reply:
Actually you are right, it doesn't make much sense to have basically two similar authentication tokens, one for the nonce the other for the session. Maybe we should just change it so that the `not_after` is only taken into account for the initial authentication and the session validity is a separate thing. And then we don't need to deal with the nonce.